File: ios/Classes/BleService.swift
import CoreBluetooth

struct BleService {
    let uuid: CBUUID
    let characteristics: [BleCharacteristic]

    static func fromMap(_ map: [String: Any]) -> BleService? {
        guard let uuidString = map["uuid"] as? String,
              let characteristicsData = map["characteristics"] as? [[String: Any]] else {
            return nil
        }
        let uuid = CBUUID(string: uuidString)

        let characteristics = characteristicsData.compactMap { BleCharacteristic.fromMap($0) }
        return BleService(uuid: uuid, characteristics: characteristics)
    }

    func toMutableService() -> CBMutableService {
        let mutableCharacteristics = characteristics.map { $0.toMutableCharacteristic() }
        let mutableService = CBMutableService(type: uuid, primary: true)
        mutableService.characteristics = mutableCharacteristics
        return mutableService
    }
}

File: android/src/main/kotlin/com/regal/flutter_ble_peripheral/BleService.kt
package com.regal.flutter_ble_peripheral

import android.bluetooth.BluetoothGattCharacteristic
import android.bluetooth.BluetoothGattService
import java.util.UUID

data class BleService(
    val uuid: UUID,
    val characteristics: List<BleCharacteristic>
) {
    companion object {
        fun fromMap(map: Map<String, Any>): BleService? {
            val uuidString = map["uuid"] as? String
            val characteristicsData = map["characteristics"] as? List<Map<String, Any>>
            if (uuidString != null && characteristicsData != null) {
                val uuid = UUID.fromString(uuidString)
                val characteristics = characteristicsData.mapNotNull { BleCharacteristic.fromMap(it) }
                return BleService(uuid, characteristics)
            }
            return null
        }
    }

    fun toGattService(): BluetoothGattService {
        val gattService = BluetoothGattService(uuid, BluetoothGattService.SERVICE_TYPE_PRIMARY)
        characteristics.forEach { characteristic ->
            gattService.addCharacteristic(characteristic.toGattCharacteristic())
        }
        return gattService
    }
}
